docs/design.md

```diff
@@ -0,0 +1,199 @@
+
+Both server and client will support TLS versions 1.2 and 1.3.
+
+The [grpc-go](https://github.com/grpc/grpc-go) package that we will be using to implement the server and client
+give direct access to the standard library's `tls.Config` struct so it should be relatively simple to verify both
+client and server certificates.
+
+### Cipher Suites
+
+#### TLS 1.2
+
+Following the guidance from these RFCs:
+
+[Recommendations for Secure Use of Transport Layer Security and Datagram Transport Layer Security (DTLS)](https://www.ietf.org/archive/id/draft-ietf-uta-rfc7525bis-03.html)
+
+[Hypertext Transfer Protocol Version 2 (HTTP/2) Appendix A](https://datatracker.ietf.org/doc/html/rfc7540#appendix-A)
+
+We will only accept cipher suites that match the following criteria:
+* Key exchange algorithm supports perfect forward secrecy
+* No RC4 cipher suites
+* No cipher suites including algorithms offering less than 128 bits of security
+* No cipher suites which use static RSA i.e cipher suites starting with TLS_RSA_WITH_*
+* Cipher Suite is not blacklisted by HTTP/2
+
+This leaves us with the following cipher suites supported by the Go TLS package https://pkg.go.dev/crypto/tls :
+
+TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
+TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
+TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
+TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
+TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256
+TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256
+
+#### TLS 1.3
+
+TLS 1.3 offers enhanced security over its predecessor, all its supported cipher suites can be negotiated.
+
+### Authorisation
+
+For authorisation, a simple all-or-nothing mechanism will be used. The server can be started with a list of OS users that are allowed to make requests.
+
+X.509 certificates issued to users will have information about the user name encoded in them in the Common Name subject field. The server will then use this information to decide if a requester has been given permission to make requests.
+
+For example, using the snippet below the users `hashi`, `mozart` and `simba` would be able to make requests to the GRPC API.
+
+```
+./worker --allowedUsers=hashi,mozart,simba
+```
+
+## Worker Library
+
+The service will include a worker library written in Go to manage and interact with Linux and Linux processes.
+Under the hood, this is what the GRPC server will use to fulfill requests.
+
+This design doc will not go into detail about the specific interfaces exposed by this library.
+
+## Process Life Cycle
+
+When there is a request to start a new process, the following will happen:
+
+* The library will fork/exec a new process (via [Cmd.Start](https://pkg.go.dev/os/exec#Cmd.Start))
+* A new cgroup will be created for this process 
+* The process PID will be written to the relevant `cgroup.procs` file
+
+When there is a request to stop a new process, the following will happen:
+* The process will be killed using https://pkg.go.dev/os#Process (we can retrieve the Process type via ProcessState)
```

Comment:
I assume you mean a kill signal, is that correct?

Reply:
Yes, I actually meant to link to this https://pkg.go.dev/os#Process.Kill but perhaps I should be talking in terms of OS instead of in terms of Go in this section.

Reply:
That is clearer, thanks.